### PR TITLE
Harmonize launching Questionnaire

### LINKED ISSUE #2180: Harmonize launching Questionnaire
**Describe the feature request.**
Use the same Questionnaire mechanism that is used by ProfileViewModel in other code places that launch a Questionnaire.

**Additional context**
The new Questionnaire mechanism is to:
1. Maps QR from DB, or else
2. Maps QR from population
3. Before launching Questionnaire, do validation for Questionnaire and QR

**Acceptance criteria**
- [ ] Works for all Questionnaires in overflow menu
  - [ ] Mapping QR from DB
  - [ ] Mapping QR from population
  - [ ] Validation for Questionnaire and QR
- [ ] Works for all Questionnaires in profile tasks
  - [ ] Mapping QR from DB
  - [ ] Mapping QR from population
  - [ ] Validation for Questionnaire and QR

### PR DESCRIPTION
**Pull request by FikriMilano**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2180

Changes:
- Refactor all Questionnaire and QuestionnaireResponse validation to QuestionnaireActivity

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [x] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [x] I have verified any strings visible on UI components are in the `strings.xml` file
- [x] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [x] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by FikriMilano** [comment]
@dubdabasoduba @ageryck kindly do a test run for all Questionnaire scenarios using this branch

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -609,6 +606,181 @@ constructor(
...
+    return try {
+      QuestionnaireResponseValidator.checkQuestionnaireResponse(
+        questionnaire,
+        questionnaireResponse
+      )
+      QuestionnaireResponseValidator.validateQuestionnaireResponse(
+        questionnaire,
+        questionnaireResponse,
+        context
+      )
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt:630]: why do we need to do this both without and with context?

**Review by pld — COMMENTED** [review]
looks good! just one question

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt (hunk 1 above)
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt:630]: Oh, it's just a parameter requirement from the SDK.
`validateQuestionnaireResponse` uses context to get a string resource for the validation results.
`checkQuestionnaireResponse ` doesn't do that so it doesn't need context.

**Review by pld — APPROVED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt [hunk 2]
```diff
@@ -146,8 +146,9 @@ fun List<Questionnaire.QuestionnaireItemComponent>.prePopulateInitialValues(
   prePopulationParams: List<ActionParameter>
 ) {
   forEach { item ->
-    if (item.hasExtension(ITEM_INITIAL_EXPRESSION_URL))
+    if (item.hasExtension(ITEM_INITIAL_EXPRESSION_URL)) {
       item.removeExtension(ITEM_INITIAL_EXPRESSION_URL)
+    }
     prePopulationParams
```
Comment (aurangzaibumer) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:152]:
```
 prePopulationParams
      .firstOrNull {
        it.linkId == item.linkId &&
          !it.value.isNullOrEmpty() &&
          !it.value.contains(interpolationPrefix)
      }
      ?.let { actionParam ->

 if (item.hasExtension(ITEM_INITIAL_EXPRESSION_URL)) {
      item.removeExtension(ITEM_INITIAL_EXPRESSION_URL)
    }

        item.initial =
          arrayListOf(
            Questionnaire.QuestionnaireItemInitialComponent().apply {
              value = actionParam.dataType?.let { actionParam.value.castToType(it) }
            }
          )
      }
```
      
      wouldn't check `removeExtension` go inside of this scope once prepopulaedPrams linkId matches with one of the questionnaire item's linkId ? 
      
      @FikriMilano

Comment (aurangzaibumer) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:152]: wouldn't check `removeExtension` go inside of this scope once prepopulaedPrams linkId matches with one of the questionnaire item's linkId ? @FikriMilano

Comment (pld) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:152]: yes that makes sense, that is unrelated to this PRs specific topic, @aurangzaibumer very good catch, I've opened an issue for this here https://github.com/opensrp/fhircore/issues/2335, please add any additional context or comment

Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:152]: @aurangzaibumer thanks, nice catch! 
Please do add the change

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt [hunk 3]
```diff
@@ -1679,4 +1653,420 @@ class QuestionnaireViewModelTest : RobolectricTest() {
...
+
+    val result = runBlocking {
+      questionnaireViewModel.getQuestionnaireResponseFromDbOrPopulation(
+        questionnaire,
+        patient.id,
+        patient.resourceType
+      )
+    }
+
+    assertTrue("Questionnaire Response has no item", result.hasItem())
```
Comment (aurangzaibumer) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt:1823]: @FikriMilano

**Pull request merged**